data-processing/common/parse_equation.py

```diff
@@ -1,151 +1,290 @@
-    # It's 'safe' to call soup.find_all as the underlying bs4 implementation
-    # appears to preserve order https://bazaar.launchpad.net/~leonardr/beautifulsoup/bs4/view/head:/bs4/element.py
-    for symbol_node in soup.find_all(SYMBOL_TAGS):
-        # Collect indexes of all characters that belong to this symbol
-        characters = []
-        if symbol_node.name in CHARACTER_TAGS:
-            index = _get_character_index(symbol_node)
-            if index is not None:
-                characters.append(index)
-        for node in symbol_node.find_all(CHARACTER_TAGS):
-            index = _get_character_index(node)
-            if index is not None:
-                characters.append(index)
-
-        # Group consecutive character symbols only if they are a child of the mrow tag.
-        # The .has_attr('s2:end') is necessary as not every symbol is annotated by katex
-        if symbol_node.has_attr('s2:end') and symbol_node.parent.name in MERGABLE_PARENT_TAGS and symbol_node.name in CHARACTER_TAGS:
-            # This symbol_node is part of consecutive_char_sequence but not the current sequence.
-            if len(consecutive_char_sequence) > 0 and not consecutive_char_sequence[-1]['s2:end'] == symbol_node['s2:start']:
-                new_symbol = _merge_consecutive_symbols(consecutive_char_sequence, consecutive_char_indices, soup.new_tag(symbol_node.name))
-                node_symbols.append(new_symbol)
-                consecutive_char_sequence = []
-                consecutive_char_indices = []
-            consecutive_char_sequence.append(symbol_node)
-            consecutive_char_indices.extend(characters)
+
+    # Parse the MathML equation, extracting top-level symbols.
+    top_level_symbols = parse_element(soup.body).symbols
+
+    # Build a list of all symbols found with a breadth-first search over the symbol tree.
+    all_symbols = []
+    symbols_to_visit = top_level_symbols
+    while len(symbols_to_visit) > 0:
+        symbol = symbols_to_visit.pop(0)
+        all_symbols.append(symbol)
+        symbols_to_visit.extend(symbol.children)
+
+    return all_symbols
+
+
+@dataclass(frozen=True)
+class ParseResult:
+
+    symbols: List[Node]
+    """
+    Top-level symbol nodes found by parsing an element. These are nodes that can be considered
+    "child symbols", if you are parsing the child tag of the parent.
+    """
+
+    element: Optional[Tag]
+    """
+    A cleaned up BeautifulSoup element of the parsed entity (i.e. with S2 metadata tags removed
+    and with consecutive identifier characters merged.). Will be None if no element could be
+    parsed from the input element.
+    """
+
+    tokens: List[Token]
+    """
+    List of all tokens found in the subtree starting at the parsed element.
+    """
+
+
+def parse_element(element: Tag) -> ParseResult:
+    """
+    Extract symbol nodes from an element in a BeautifulSoup MathML parse tree. This function
+    recursively visits
```

Comment:
nit: "This function recursively visits nodes"

Reply:
Good catch! I must have run out of steam on that comment XD Will fix it up.